### PR TITLE
Python Fixes + Header Params + Authentications with API Key + Stages + Summary + Description + Required

### PR DESCRIPTION
**Pull request by fcrdossantos**
**Features:**
- Enabled Authetication with ApiKey (Authorize)
   - Using it you can Authorize yourself in swagger to execute requests
- Enabled field Required in Parameters 
   - Fixed "Required" in QueryParams and added it on HeaderParams
- API Description and Summary
   - You can now insert a Description or Summary to each end-point
- Header Parameters
   - Now you can use Headers Params as well (not only QueryParms)
- Fixed Python Generation 
   - Now the swagger files under "Resources" are working with Python
- Stages in Path 
   - You can decide if you want or not to use the actual stage on beginning of path
- httpApi usage
   - httpApi is the new, faster and cheaper "http". So i'm using it

**README:**
All the features above are described in README.md file with usage example.

I hope this RP helps the community :)

### CODE REVIEW

File: src/resources/functions.ts
Comment (desaintvincent) [inline src/resources/functions.ts:25]: path: useStage ? `/${stage}/swagger` : `/swagger`,

that would prevent code duplication in this big if else

**Comment by desaintvincent** [comment]
have some lint rules would definitively be a plus on this project

File: src/resources/functions.ts
Comment (fcrdossantos) [reply src/resources/functions.ts:25]: Ooh, thanks for the tip :)

I was doing that on another way and when i changed it, i forgot a code review hahaha

I think there's another places that we can improve too, as JS/TS is not a language that i am used to, i think that a really have a lot to improve 

Maybe when Sam has some time he could improve my code :)

Comment (fcrdossantos) [reply src/resources/functions.ts]: @desaintvincent i submited another commit with your correction, thanks :)

**Comment by fcrdossantos** [comment]
> have some lint rules would definitively be a plus on this project

As i dont have a good knowledge in JS/TS Linting, i just ran the default ESLint on project now, i hope this could help :)

**Comment by desaintvincent** [comment]
@fcrdossantos I don't think it should be done in this MR as it is really out of scope. It was just a little thinking because I see that some of the changed lines are just about "code lint".

Some lines have `;` some don't, some lines has double space tabulation, some 4. same for quote. With multiple devs with multiple configs on the same code, I think it is better to stick to a commun "project config"

File: package.json
Comment (bfaulk96) [inline package.json:20]: `copyfiles` could be a dev dependency since it's used only for `npm run build`, not for the actual package itself, right?

File: src/serverlessPlugin.d.ts
Comment (bfaulk96) [inline src/serverlessPlugin.d.ts:167]: Was it intentional to leave this in?

File: src/swagger.d.ts
Comment (bfaulk96) [inline src/swagger.d.ts:85]: Typescript now has the `Record` utility type so you don't have to do this
```suggestion
  scopes?: Record<string, string>
```

File: package.json
Comment (fcrdossantos) [reply package.json:20]: Done, thanks :)

As i'm not a node/js developer, i just installed it normally by npm install and npm put it there. I'll move it :)

File: src/serverlessPlugin.d.ts
Comment (fcrdossantos) [reply src/serverlessPlugin.d.ts:167]: Removed, thanks.

File: src/swagger.d.ts
Comment (fcrdossantos) [reply src/swagger.d.ts:85]: Hm... This file wasn't eddited by me, i don't know why @SamWSoftware did it, but as i'm not a specialist in node/serverless-plugins and all the rest of code uses "[key: string]: string" maybe it's better to stay this way in my opnion

But thanks for the tip, i'll search about Record :)

**Comment by fcrdossantos** [comment]
I just sent a fix for Serverless 3.0

I think i'm done with the changes i need to my project to work.

If this PR wont be accepted and you guys need to use it with Serverless 3.0.0 or need it working with Python, please use my fork   :)

File: src/ServerlessAutoSwagger.ts
Comment (bfaulk96) [inline src/ServerlessAutoSwagger.ts:126]: @SamWSoftware I don't know much about this, does this section look good to you?

File: src/ServerlessAutoSwagger.ts
Comment (bfaulk96) [inline src/ServerlessAutoSwagger.ts:257]: Should be `const` since it's not redefined anywhere

File: src/ServerlessAutoSwagger.ts
Comment (bfaulk96) [inline src/ServerlessAutoSwagger.ts:268]: Another fun Typescript trick :)
```suggestion
        securityDefinitions,
```

File: src/ServerlessAutoSwagger.ts
Comment (bfaulk96) [reply src/ServerlessAutoSwagger.ts:257]:
```suggestion
    const securityDefinitions: { [key: string]: SecurityDefinition } = {}
```

File: src/ServerlessAutoSwagger.ts
Comment (bfaulk96) [inline src/ServerlessAutoSwagger.ts:351]:
```suggestion
          const security_key_name: { [key: string]: string[] } = {}
```

File: src/ServerlessAutoSwagger.ts
Comment (bfaulk96) [inline src/ServerlessAutoSwagger.ts:354]:
```suggestion
          var securityKey = [securityKeyName] as MethodSecurity[]
```

File: src/ServerlessAutoSwagger.ts
Comment (bfaulk96) [inline src/ServerlessAutoSwagger.ts:365]: @SamWSoftware Again, I don't want to mark this as good without fully understnading, is `http.summary` fine? should we default `?? functionName` here?

File: src/ServerlessAutoSwagger.ts
Comment (bfaulk96) [inline src/ServerlessAutoSwagger.ts:366]: I'm thinking `functionName` should be the fallback for `summary` instead of description, but I'm curious to hear @SamWSoftware 's opinion on it

**Review by bfaulk96 — COMMENTED** [review]
From what I can tell, the headerParameters stuff looks fine other than the fact that it's missing unit testing... Any chance you could throw some quick unit tests in there for it? Should be pretty straightforward since headerParamters seem to be handled similarly to queryParameters.

**Comment by bfaulk96** [comment]
I've fixed these merge conflicts, as well as a majority of the PR comments I left, but I am unable to merge to your fork. As I'm fairly new to open-source contribution I'm not really sure how to share my changes with you without creating a new branch entirely.

**Comment by bfaulk96** [comment]
I created `fcrdossantos-patch-1` branch if you'd like to pull that into your branch to fix these conflicts.
I'd still like @SamWSoftware's opinion on the `functionName` in the summary/description, but other than that I think I addressed the few small comments I made on your PR.

**Comment by bfaulk96** [comment]
Actually upon a second look at this, this PR breaks all sorts of existing unit tests as well... Looks like a lot of undefined values are set when they probably shouldn't be

File: README.md
Comment (bfaulk96) [inline README.md:19]: Just a heads up, I'm going to remove this on my fork of your branch. I'll add them to the CHANGELOG when creating another version to publish.

**Comment by bfaulk96** [comment]
Since I asked a whole lot out of you (and you've already mentioned you don't really use TS), I made another PR forking yours and making the above mentioned changes. https://github.com/SamWSoftware/serverless-auto-swagger/pull/22

**Comment by fcrdossantos** [comment]
Wow! Thanks a lot @bfaulk96  :) 

I'll check your PR and look all the changes

Your PR helps a lot, so we ensure that the code is better structured and following the TS guidelines.

About the changes requested above, if I'm not mistaken, the Summary takes the name of the function by default if there is no value.

When you "maximize" a route in swagger, the first line of that block is the summary, at least in my tests, if I didn't type anything it got the function name by default.

However, for safety's sake, we can already make this explicit in TS just like you recommended :)

**Comment by bfaulk96** [comment]
> When you "maximize" a route in swagger, the first line of that block is the summary, at least in my tests, if I didn't type anything it got the function name by default.

> However, for safety's sake, we can already make this explicit in TS just like you recommended :)

Alright! I'll update the unit tests, thanks for the explanation!

**Comment by fcrdossantos** [comment]
Hi @bfaulk96, if i can ask you one more thing, i noticed that i will need to use multi-valued query parameter on one of my APIs.

Basically we need to tell Swagger that the query type is `arry` and then add 2 properties:

```
items: 
 - type: string | integer
```

and 

```
collectionFormat: multi
```

I merged your changes into my fork and then implemented this feature. If you can take a look :)


Edit: I noticed that my VS Code removed ";" and break few lines due to an auto-formatter that i use on Python (black) 
If you want, you can edit/modify it to stay on your Lint settings :) 

Thanks again for your time and help

File: src/ServerlessAutoSwagger.ts
Comment (bfaulk96) [inline src/ServerlessAutoSwagger.ts:489]:
```suggestion
          ...(data.type === "array" ? { items: { type: data.itemsType }, collectionFormat: "multi" } : {}),
```

File: src/serverlessPlugin.d.ts
Comment (bfaulk96) [inline src/serverlessPlugin.d.ts:100]: This should be more explicit like `arrayItemsType` to be clear that it's specific to the array type

Comment (fcrdossantos) [reply src/serverlessPlugin.d.ts:100]: True! I put it on README only, but your name is better :) 

I'll change it and commit again, thanks

**Review by bfaulk96 — CHANGES_REQUESTED** [review]
You are once again missing unit test for the new feature you're trying to add. After you add unit tests for this, please do not add any more features to this pull request. In general, pull requests should be limited to one feature or bug, not a slew of random features, as it can lead to long reviews back and forth like this.

It's super awesome that you're so thrilled to contribute these great changes, I'm just letting you know it makes it significantly harder for both myself and you when there's an ever-growing list of features thrown into one PR 😄

**Comment by bfaulk96** [comment]
No hard feelings (I understand why you wanted to, since there was not much styling consistency in this repo), but I really wish you hadn't run your linter on this PR at all - since there are already so many features and changes it made it much more difficult to look through and review. It's for that same reason I chose not to fix the linting until after this PR is merged; it really should be its own PR.

File: README.md
Comment (bfaulk96) [inline README.md:139]: I think this is my last comment to address other than those unit tests, sorry I missed this in my review a minute ago
```suggestion
If you use multi-value query string parameters (array), then you must specify that your `type` is `array` and specify your data type (string or integer) in `arrayItemsType`

```

**Comment by fcrdossantos** [comment]
Sure thanks, the PR was increased as i noticed soma remaining features, but after that one, all my APIS are working and i thing that i just finished with it :) 

I hope that on the future @SamWSoftware  can take a look and implement some of then. I Hope it helps some Python devs :)

**Comment by fcrdossantos** [comment]
Done, added the unit test

I think i done with it now :) 

Thanks for your help until here, i hope those fix/features helps some Python devs :) 

Also, if you want, you can close this PR, your branch has the fixes, i think it's enough to Sam to merge if in the future

**Review by bfaulk96 — APPROVED** [review]

**Comment by bfaulk96** [comment]
Looks good, I'll merge this as soon as I can get ahold of Sam to set up publishing to NPM (The previous PR that was merged has still not been deployed, so I'd like to deploy that one and then tag a different version for these changes)

**Comment by bfaulk96** [comment]
> Also, if you want, you can close this PR, your branch has the fixes, i think it's enough to Sam to merge if in the future

And my PR doesn't have your multi- query params updates, so I'll just merge this one when I can and close mine 😃

File: src/ServerlessAutoSwagger.ts
Comment (SamWSoftware) [reply src/ServerlessAutoSwagger.ts]: I struggle to understand what the difference between description and summary is. Either would be fine in my opinion.

File: src/ServerlessAutoSwagger.ts
Comment (SamWSoftware) [reply src/ServerlessAutoSwagger.ts]: summary and description are optional so no need to default them both.

**Comment by SamWSoftware** [comment]
Features:

- [ ] Enabled Authetication with ApiKey (Authorize)
- [x] Enabled field Required in Parameters
- [x] API Description and Summary
- [X] Header Parameters
- [x] Fixed Python Generation
- [ ] Stages in Path - I cannot find 'useStage' anywhere in the code. Was this removed?
- [x] httpApi usage

**Pull request merged**
